clang-tools-extra/clang-tidy/bugprone/InvalidEnumDefaultInitializationCheck.cpp

```diff
@@ -0,0 +1,182 @@
+  }
+  bool VisitConstantArrayType(const ConstantArrayType *T) {
+    return Visit(T->getElementType()->getUnqualifiedDesugaredType());
+  }
+  bool VisitEnumType(const EnumType *T) {
+    if (isCompleteAndHasNoZeroValue(T->getDecl())) {
+      FoundEnum = T;
+      return true;
+    }
+    return false;
+  }
+  bool VisitRecordType(const RecordType *T) {
+    const RecordDecl *RD = T->getDecl();
+    if (RD->isUnion())
+      return false;
+    auto VisitField = [this](const FieldDecl *F) {
+      return Visit(F->getType()->getUnqualifiedDesugaredType());
+    };
+    return llvm::any_of(RD->fields(), VisitField);
+  }
+};
+
+} // namespace
+
+InvalidEnumDefaultInitializationCheck::InvalidEnumDefaultInitializationCheck(
+    StringRef Name, ClangTidyContext *Context)
+    : ClangTidyCheck(Name, Context) {}
+
+void InvalidEnumDefaultInitializationCheck::registerMatchers(
+    MatchFinder *Finder) {
+  auto EnumWithoutZeroValue = enumType(
+      hasDeclaration(enumDecl(isCompleteAndHasNoZeroValue()).bind("enum")));
+  auto EnumOrArrayOfEnum = qualType(hasUnqualifiedDesugaredType(
+      anyOf(EnumWithoutZeroValue,
+            arrayType(hasElementType(qualType(
+                hasUnqualifiedDesugaredType(EnumWithoutZeroValue)))))));
+  Finder->addMatcher(
+      expr(isEmptyInit(), hasType(EnumOrArrayOfEnum)).bind("expr"), this);
+
+  // Array initialization can contain an "array filler" for the (syntactically)
+  // unspecified elements. This expression is not found by AST matchers and can
+  // have any type (the array's element type). This is an implicitly generated
+  // initialization, so if the type contains somewhere an enum without zero
+  // enumerator, the zero initialization applies here. We search this array
+  // element type for the specific enum type manually when this matcher matches.
+  Finder->addMatcher(initListExpr(hasArrayFiller()).bind("array_filler_expr"),
+                     this);
+}
+
+void InvalidEnumDefaultInitializationCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *InitExpr = Result.Nodes.getNodeAs<Expr>("expr");
+  const auto *Enum = Result.Nodes.getNodeAs<EnumDecl>("enum");
+  if (!InitExpr) {
+    const auto *InitList =
+        Result.Nodes.getNodeAs<InitListExpr>("array_filler_expr");
+    // Initialization of omitted array elements with array filler was found.
+    // Check the type for enum without zero value.
+    // FIXME: In this way only one enum-typed value is found, not all of these.
+    FindEnumMember Finder;
+    if (!Finder.Visit(InitList->getArrayFiller()->getType().getTypePtr()))
+      return;
+    InitExpr = InitList;
+    Enum = Finder.FoundEnum->getDecl();
+  }
+
+  if (!InitExpr || !Enum)
+    return;
+
+  ASTContext &ACtx = Enum->getASTContext();
+  SourceLocation Loc = InitExpr->getExprLoc();
+  if (Loc.isInvalid()) {
+    if (isa<ImplicitValueInitExpr>(InitExpr) || isa<InitListExpr>(InitExpr)) {
```

Comment:
I think isa<>() now supports variadic args, so we could use:
isa<ImplicitValueInitExpr, InitListExpr>(InitExpr)